eth_account/hdaccount/mnemonic.py

```diff
@@ -0,0 +1,161 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+
+import binascii
+import bisect
+import hashlib
+import os
+import unicodedata
+
+PBKDF2_ROUNDS = 2048
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+# https://stackoverflow.com/questions/212358/binary-search-bisection-in-python/2233940#2233940
+def binary_search(a, x, lo=0, hi=None):  # can't use a to specify default for hi
+    hi = hi if hi is not None else len(a)  # hi defaults to len(a)
+    pos = bisect.bisect_left(a, x, lo, hi)  # find insertion position
+    return pos if pos != hi and a[pos] == x else -1  # don't walk off the end
+
+
+class Mnemonic(object):
+    def __init__(self, language):
+        self.radix = 2048
+        with open(f"{self._get_directory()}/{language}.txt", "r", encoding="utf-8") as f:
+            self.wordlist = [w.strip() for w in f.readlines()]
+        if len(self.wordlist) != self.radix:
+            raise ConfigurationError(
+                f"Wordlist should contain {self.radix} words, "
+                f"but it contains {len(self.wordlist)} words."
+            )
+
+    @classmethod
+    def _get_directory(cls):
+        return os.path.join(os.path.dirname(__file__), "wordlist")
+
+    @classmethod
+    def list_languages(cls):
+        return [
+            f.split(".")[0]
+            for f in os.listdir(cls._get_directory())
+            if f.endswith(".txt")
+        ]
+
+    @classmethod
+    def normalize_string(cls, txt):
+        if isinstance(txt, bytes):
+            utxt = txt.decode("utf8")
+        elif isinstance(txt, str):
+            utxt = txt
+        else:
+            raise TypeError("String value expected")
+
+        return unicodedata.normalize("NFKD", utxt)
+
+    @classmethod
+    def detect_language(cls, code):
+        code = cls.normalize_string(code)
+        first = code.split(" ")[0]
+        languages = cls.list_languages()
+
+        for lang in languages:
+            mnemo = cls(lang)
+            if first in mnemo.wordlist:
+                return lang
+
+        raise ConfigurationError("Language not detected")
+
+    def generate(self, strength=128):
+        if strength not in [128, 160, 192, 224, 256]:
+            raise ValueError(
+                "Strength should be one of the following: [128, 160, 192, 224, 256]"
+            )
+        return self.to_mnemonic(os.urandom(strength // 8))
+
+    def to_mnemonic(self, data):
+        if len(data) not in [16, 20, 24, 28, 32]:
+            raise ValueError(
+                "Data length should be one of the following: [16, 20, 24, 28, 32]"
+                % len(data)
+            )
+        h = hashlib.sha256(data).hexdigest()
+        b = (
+            bin(int(binascii.hexlify(data), 16))[2:].zfill(len(data) * 8) +
```

Comment:
This library I more or less copied wholesale, and removed unused parts, so we can do this in a clearer way.

This module was the reference implemention for BIP39, so I figured that it could be considered of higher pedigree.

Reply:
Yeah, for sure. I'm not suggesting a change, just wondering if there was something I didn't see

Reply:
I fixed this up a lot to make it clearer, but basically there are 11 bits per word choice from the list (2**11 = 2048), so this is their way of trying to obtain 11 bit chunks from the raw seed